loopy/target/c/__init__.py

```diff
@@ -1123,4 +1156,105 @@ def get_host_ast_builder(self):
+        super(DTypeRegistryWrapperVec, self).__init__(wrapped_registry)
+
+    def dtype_to_ctype(self, dtype):
+        base = self.dtype_to_ctype_base(dtype)
+        # two worded typed need to be merged to one word
+        # so that vector type can be defined afterwards
+        if len(base.split()) > 1:
+            base = "".join([word[0] for word in base.split()])
+        if dtype.dtype.shape:
+            shape, = dtype.dtype.shape
+            return base + str(shape)
+        return base
+    
+    def dtype_to_ctype_base(self, dtype):
+        return super(DTypeRegistryWrapperVec, self).dtype_to_ctype(
+                NumpyType(dtype.dtype.base))
+
+
+class CVecASTBuilder(CASTBuilder):
+
+    def preamble_generators(self):
+        obj = [
+                cvec_preamble_generator,
+              ] + super(CVecASTBuilder, self).preamble_generators()
+        return (obj)
+
+    def get_expression_to_c_expression_mapper(self, codegen_state):
+        from loopy.target.c.codegen.expression import ExpressionToCVecExpressionMapper  # noqa
+        return ExpressionToCVecExpressionMapper(codegen_state)
+
+    def get_c_expression_to_code_mapper(self):
+        from loopy.target.c.codegen.expression import CVecExpressionToCodeMapper
+        return CVecExpressionToCodeMapper()
+
+    def emit_sequential_loop(self, codegen_state, iname, iname_dtype,
+            lbound, ubound, inner):
+
+        from loopy.kernel.data import CVectorizeTag, OpenMPSIMDTag
+
+        if codegen_state.kernel.iname_tags_of_type(iname, CVectorizeTag):
+            return inner
+
+        if codegen_state.kernel.iname_tags_of_type(iname, OpenMPSIMDTag):
+            from cgen import Block, Pragma
+            loop = super(CVecASTBuilder, self).emit_sequential_loop(
+                codegen_state, iname, iname_dtype, lbound, ubound, inner)
+            loop = Block([Pragma("omp simd"), loop])
+            return loop
+
+        return super(CVecASTBuilder, self).emit_sequential_loop(
+            codegen_state, iname, iname_dtype, lbound, ubound, inner)
+
+    def get_temporary_decl(self, codegen_state, schedule_index, temp_var, decl_info):
+        if hasattr(temp_var, "zero_size"):
+            if temp_var.zero_size > 1:
+                temp_var_decl = POD(self, decl_info.dtype, decl_info.name,
+                                    temp_var.zero_size)
+
+                from cgen import Const
+                temp_var_decl = Const(temp_var_decl)
+
+                if temp_var.alignment:
+                    from cgen import AlignedAttribute
+                    temp_var_decl = AlignedAttribute(temp_var.alignment, temp_var_decl)
+
+                return temp_var_decl
+
+        return super(CVecASTBuilder, self).get_temporary_decl(
+            codegen_state, schedule_index, temp_var, decl_info)
+
+
+class CVecTarget(CTarget):
+    """C Target with vector extensions, e.g. double4
+    """
+
+    def __init__(self, length):
```

Comment:
Compared to TJ I have added that the CVecTarget gets initialised with the batch size. The information is necessary to generate the right preamble, e.g. to typedef double to double4. For an instruction with a constant right hand side, we need to add a vector type zero otherwise the compiler won't vectorise - for the declaration of those zeros the batch size is also necessary.